nibabel/streamlines/tractogram.py

```diff
@@ -0,0 +1,543 @@
+            points for a particular streamline t and M is the number of
+            scalars associated to each point (excluding the three
+            coordinates).
+        """
+        self.streamlines = streamlines
+        self.data_per_streamline = data_per_streamline
+        self.data_per_point = data_per_point
+        self._affine_to_rasmm = np.eye(4)
+
+    @property
+    def streamlines(self):
+        return self._streamlines
+
+    @streamlines.setter
+    def streamlines(self, value):
+        self._streamlines = ArraySequence(value)
+
+    @property
+    def data_per_streamline(self):
+        return self._data_per_streamline
+
+    @data_per_streamline.setter
+    def data_per_streamline(self, value):
+        self._data_per_streamline = Tractogram.DataPerStreamlineDict(self,
+                                                                     value)
+
+    @property
+    def data_per_point(self):
+        return self._data_per_point
+
+    @data_per_point.setter
+    def data_per_point(self, value):
+        self._data_per_point = Tractogram.DataPerPointDict(self, value)
+
+    @property
+    def affine_to_rasmm(self):
+        # Return a copy. User should use self.apply_affine` to modify it.
+        return self._affine_to_rasmm.copy()
+
+    def __iter__(self):
+        for i in range(len(self.streamlines)):
+            yield self[i]
+
+    def __getitem__(self, idx):
+        pts = self.streamlines[idx]
+
+        data_per_streamline = {}
+        for key in self.data_per_streamline:
+            data_per_streamline[key] = self.data_per_streamline[key][idx]
+
+        data_per_point = {}
+        for key in self.data_per_point:
+            data_per_point[key] = self.data_per_point[key][idx]
+
+        if isinstance(idx, (int, np.integer)):
+            return TractogramItem(pts, data_per_streamline, data_per_point)
+
+        return Tractogram(pts, data_per_streamline, data_per_point)
+
+    def __len__(self):
+        return len(self.streamlines)
+
+    def copy(self):
+        """ Returns a copy of this :class:`Tractogram` object. """
+        data_per_streamline = {}
+        for key in self.data_per_streamline:
+            data_per_streamline[key] = self.data_per_streamline[key].copy()
+
+        data_per_point = {}
+        for key in self.data_per_point:
+                data_per_point[key] = self.data_per_point[key].copy()
+
+        tractogram = Tractogram(self.streamlines.copy(),
+                                data_per_streamline,
+                                data_per_point)
+
+        tractogram._affine_to_rasmm = self.affine_to_rasmm
+        return tractogram
+
+    def apply_affine(self, affine, lazy=False):
+        """ Applies an affine transformation on the points of each streamline.
+
+        If `lazy` is not specified, this is performed *in-place*.
+
+        Parameters
+        ----------
+        affine : ndarray shape (4, 4)
```

Comment:
I'm sorry - it's getting a little late here, and I'm confused.

The points are defined as being, before this routine gets called, in RAS+ space.

Then we apply the `affine` to the points.

What space are they in now?  I'm guessing RAS+ space - i.e. we've changed our mind about how to get to RAS+ space, and this affine is our way of getting to the new, real RAS+ space.

So, I would imagine that `self._affine_to_rasmm` would be ``affine.dot(self._affine_to_rasmm)`` - but I see that's not so.  Could you explain?